tensorflow/lite/micro/kernels/reverse.cc

```diff
@@ -0,0 +1,201 @@
+
+  if (input->type != kTfLiteInt32 && input->type != kTfLiteFloat32 &&
+      input->type != kTfLiteUInt8 && input->type != kTfLiteInt8 &&
+      input->type != kTfLiteInt16) {
+    MicroPrintf("Type '%s' is not supported by reverse.",
+                TfLiteTypeGetName(input->type));
+    return kTfLiteError;
+  }
+
+  if (axis->type != kTfLiteInt32) {
+    MicroPrintf("Axis Type '%s' is not supported by reverse.",
+                TfLiteTypeGetName(axis->type));
+    return kTfLiteError;
+  }
+  // The value type and output type must match.
+  TF_LITE_ENSURE_EQ(context, input->type, output->type);
+
+  micro_context->DeallocateTempTfLiteTensor(input);
+  micro_context->DeallocateTempTfLiteTensor(axis);
+  micro_context->DeallocateTempTfLiteTensor(output);
+  return kTfLiteOk;
+}
+
+template <typename T>
+void ReverseImpl(int32_t* axes, int num_axes, const RuntimeShape& input_shape,
+                 const T* input_data, T* output_data) {
+  bool is_upper = (axes[num_axes - 1] == input_shape.DimensionsCount() - 1);
+  bool is_lower = (axes[0] == 0);
+  int rank = input_shape.DimensionsCount();
+  if (is_upper && is_lower) {
+    std::reverse_copy(input_data, input_data + input_shape.FlatSize(),
+                      output_data);
+    return;
+  } else {
+    int32_t min_dim = axes[0];
+    int32_t max_dim = axes[num_axes - 1];
+    int upper_size = 1;
+    for (int i = 0; i < min_dim; ++i) {
+      upper_size *= input_shape.Dims(i);
+    }
+    int lower_size = 1;
+    for (int i = max_dim + 1; i < rank; ++i) {
+      lower_size *= input_shape.Dims(i);
+    }
+    int middle_size = 1;
+    for (int i = min_dim; i <= max_dim; ++i) {
+      middle_size *= input_shape.Dims(i);
+    }
+
+    if (lower_size > 1) {
+      for (int i = 0; i < upper_size; ++i) {
+        for (int j = 0; j < middle_size; ++j) {
+          T* src = (T*)input_data + (i * (middle_size) + j) * lower_size;
+          T* dst = (T*)output_data +
+                   (i * (middle_size) + (middle_size - j - 1)) * lower_size;
+          memcpy(dst, src, lower_size * sizeof(T));
+        }
+      }
+    } else {
+      for (int i = 0; i < upper_size; ++i) {
+        std::reverse_copy(input_data + i * (middle_size),
+                          input_data + i * middle_size + middle_size,
+                          output_data + i * (middle_size));
+      }
+    }
+  }
+}
+
+TfLiteStatus ReverseV2Eval(TfLiteContext* context, TfLiteNode* node) {
+  const TfLiteEvalTensor* input =
+      micro::GetEvalInput(context, node, kInputTensor);
+  const TfLiteEvalTensor* axis =
+      micro::GetEvalInput(context, node, kAxisTensor);
+  TfLiteEvalTensor* output = micro::GetEvalOutput(context, node, kOutputTensor);
+
+  TF_LITE_ENSURE_EQ(context, axis->type, kTfLiteInt32);
+  const int num_axes = static_cast<int>(ElementCount(*axis->dims));
+  TF_LITE_ENSURE(context, num_axes <= 8);
+
+  int32_t axes_data[kMaxDimensions];
+  std::memcpy(axes_data, axis->data.i32, sizeof(int32_t) * num_axes);
```

Comment:
This should be `axis->data.data`.  Use of any other member of the union is deprecated.